docs/testing-with-phpunit.md

```diff
@@ -0,0 +1,225 @@
+
+Within the `.tests` directory you can organise your test classes as you see fit however it is good practice to group related tests into directories. This makes it easier to run a subset of tests rather than the full suite.
+
+Group tests into a class when they test different aspects of the same piece of functionality. It’s especially convenient to put tests together in a class when they can share a common `setUp()` or `setUpBeforeClass()` routine. As a rule, a single test class should not contain tests for more than one function/method and should test every possible input and expected output.
+
+### Unit Tests
+
+To create a unit test your class must extend the `PHPUnit\FrameWork\TestCase` class and the methods that contain [assertions](https://phpunit.readthedocs.io/en/7.1/assertions.html) must start with `test`.
+
+```php
+<?php
+
+namespace Project\Tests;
+
+use PHPUnit\Framework\TestCase;
+
+class Test_Units extends TestCase {
+
+	/**
+	 * This special method is run once before all the tests in the class.
+	 */
+	public function setUpBeforeClass() {
+		require_once 'path/to/file/with/functions.php';
+	}
+
+	public function test_complex_maths_function() {
+		// Run the function to test.
+		$value = complex_maths( 22.24, 87 );
+		// Make sure the value is what we expect.
+		$this->assertSame( $value, 7 );
+	}
+
+}
+```
+
+The framework has a great many features, it is highly recommended to [read through the PHPUnit documentation](https://phpunit.readthedocs.io/en/7.1/writing-tests-for-phpunit.html) to get the most value out of it.
+
+### Integration Tests
+
+To run tests against the running application your tests should follow the same pattern as above but test classes should extend the `WP_UnitTestCase` class.
+
+```php
+<?php
+
+namespace Project\Tests;
+
+use WP_UnitTestCase;
+
+class Test_Custom_Post_Type extends WP_UnitTestCase {
+
+	public function test_post_type_exists() {
+		// The full application and it's functions are loaded.
+		$this->assertTrue( post_type_exists( 'event' ) );
+	}
+
+}
+```
+
+The `WP_UnitTestCase` class provides a factory object for creating content, users and more. This factory object handles resetting data so you can be confident individual tests are running in a consistent environment.
+
+The following example demonstrates adding a user to be persisted across all tests and a test that creates a post with that user as the author. The user will exist for all tests in the class while the post created within the test method will be cleaned up immediately after that test.
+
+```php
+<?php
+
+namespace Project\Tests;
+
+use WP_UnitTestCase;
+
+class Test_Custom_Post_Type extends WP_UnitTestCase {
+
+	/**
+	 * Store the user ID.
+	 *
+	 * @var int
+	 */
+	protected static $user_id;
+
+	/**
+	 * Runs before the entire class and recieves the WP factory object.
+	 */
+	public function wpSetUpBeforeClass( $factory ) {
```

Comment:
Everything created in `wpSetUpBeforeClass()` needs to be deleted in `wpTearDownAfterClass()`.

Reply:
Right, I thought that was handled automatically. Cheers

Reply:
Added a corresponding tear down method.